src/metabase/models/field_values.clj

```diff
@@ -321,35 +341,72 @@
 ;;; |                                                    CRUD fns                                                    |
 ;;; +----------------------------------------------------------------------------------------------------------------+
 
+(defn- distinct-text-field-rff
+  "Similar to [[metabase.query-processor.reducible/default-rff]] but stop when the total characters of all rows exceed
+  `max-char-len`.
+
+  Expect the row to have only one column of text."
+  [max-char-len]
+  (fn [metadata]
+   (let [row-count  (volatile! 0)
+         rows       (volatile! (transient []))
+         total-char (volatile! 0)
+         offer-row  (fn [row]
+                     (vswap! row-count inc)
+                     (vswap! rows conj! row))]
+     (fn default-rf
+       ([]
+        {:data metadata})
+
+       ([result]
+        {:pre [(map? (unreduced result))]}
+        ;; if the result is a clojure.lang.Reduced, unwrap it so we always get back the standard-format map
+        (-> (unreduced result)
+            (assoc :row_count @row-count
+                   :status :completed)
+            (assoc-in [:data :rows] (persistent! @rows))))
+
+       ([result row]
+        (assert (= 1 (count row)))
+        (vswap! total-char + (count (first row)))
+        (if (> @total-char max-char-len)
+          (reduced (assoc result ::reached-char-len-limit true))
+          (do
+           (offer-row row)
+           result)))))))
+
 (defn distinct-values
   "Fetch a sequence of distinct values for `field` that are below the [[*total-max-length*]] threshold. If the values are
   past the threshold, this returns a subset of possible values values where the total length of all items is less than [[*total-max-length*]].
   It also returns a `has_more_values` flag, `has_more_values` = `true` when the returned values list is a subset of all possible values.
 
   ;; (distinct-values (Field 1))
   ;; ->  {:values          [[1], [2], [3]]
-          :has_more_values false}
+  :has_more_values false}
 
   (This function provides the values that normally get saved as a Field's
   FieldValues. You most likely should not be using this directly in code outside of this namespace, unless it's for a
   very specific reason, such as certain cases where we fetch ad-hoc FieldValues for GTAP-filtered Fields.)"
   [field]
   (try
-    (let [distinct-values         (metadata-queries/field-distinct-values field)
-          limited-distinct-values (take-by-length *total-max-length* distinct-values)]
-      {:values          limited-distinct-values
+    (let [rff             (if (isa? (:base_type field) :type/Text)
```

Comment:
Why is this only applied to `:type/Text` fields? `take-by-length` applied to all types and just assumed that all values had a `.toString()` method.

Reply:
for reference, the source of `take-by-length`:
https://github.com/metabase/metabase/blob/49654444c9ae810e31f85a94e1f38538307d92fe/src/metabase/models/field_values.clj#L260-L279

Reply:
I wasn't sure it's useful for other types, but I guess you're right.

Reply:
Do you think it's useful for types other than text? I can see float number being extra long but it's unlikely

Reply:
it's a good question. doing this for other types is probably not useful right now. on the other hand there could be other types in the future that are long that aren't `:type/Text` and are `:type/TextLike` or something. doing this on every value is rather cheap though, since we are converting to strings anyway

Reply:
make sense, did it in [4e6f6e6](https://github.com/metabase/metabase/pull/46524/commits/4e6f6e6de480a6328af8a7d60f035d65ce8ba3c5)